### ISSUE TITLE
Evaluation in robomimic

### ISSUE THREAD

**Issue by MarSaKi** [Max-Fu/icrt#8]
Hi, thanks for your great work and generous open-sourcing effort! I really appreciate ICRT and I think it point out another promising way for robot learning rather than VLA.

I noticed you have open-sourced the [code](https://github.com/Max-Fu/icrt/blob/main/scripts/eval_fk.py) for evaluating ICRT in robomimic, while I found this part is not complete, such as the file ["env_attr_square_lift_can.json"](https://github.com/Max-Fu/icrt/blob/fd5f2fb1305f466b55f7aa49cf1b0109fa832adc/scripts/eval_fk.py#L477) is missing. 

I would like to kindly inquire how to evaluate ICRT in simulation such as robomimic or robocasa, and do you have plan to open the code of simulation evaluation? If not, I'd like to describe my expected process and could you please give me some feedback?
1. Collect or download data from robomimic, and transfer them as the droid data format using the [scripts](https://github.com/Max-Fu/icrt/blob/main/DATASET.md) you provided. 
2. Initialize the model with DROID pretrained parameters and fine-tune it using the dataset in step 1. 
3. Evaluate the policy rollout using the model trained in step 2. 

Besides, for eval_fk.py, which version of robomimic did you use and can it support robocasa? Do you have recommended simulation or benchmarks to test the ICRT algorithm (or the camera, data steup is aligned with ICRT)? 

Thanks for your attention to this matter!

**Comment by Max-Fu** [comment]
Hi, 

Thanks for your interest in this work! I think for the time being, I probably don't have the bandwidth to eval ICRT on robomimic/robocasa. I think your expected process is correct. The problem I had with simulation was that it was hard to tell if a model is actually instruction following or it is just overfitting to the training tasks. In addition, policies that worked in sim didn't work in real (we ended up finding that action chunking is important). All these factors led to no sim eval for this paper. To evaluate instruction following, changes needs to be made to the environment config (i.e. change the color of the target objects, add new distractors than the ones that the model has seen, etc.) I think LIBERO is a good simulator where these configs are modifiable. My collaborator on [OTTER](https://ottervla.github.io/) will release the environment configs for testing so look out for that!

Best,
Max

**Comment by MarSaKi** [comment]
Hi, Max, thanks for your quick reply and sharing the valuable real-robot experience. I really appreciate it!

I still have some questions about the mechanism of ICRT, and may I have your thoughts about it? 😶‍🌫️
According to this [paper](https://arxiv.org/pdf/2305.09731), there are probably two mechanism behind ICL:
1. Task Recognition (TR): using the demonstration to identify the task seen during pretraining. 
2. Task Learning (TL): learning new transformation from demonstration. And I think this is akin to your statement in Section "Data Formulation": 
![Image](https://github.com/user-attachments/assets/2018dc98-b02c-4026-93d7-e2ac4155093d)

I'm wondering about the role of the prompt trajectory in ICRT. As I understand it, a prompt trajectory not only indicates **which object** to interact with, but also **how to execute** a particular action primitive.

In the _Problem Statement section_, you define new tasks as those involving unseen objects, but using motion primitives from the training data. This suggests that the prompt trajectory is used to specify the **target object, the placement, and which action primitive to execute**. However, **the way the action primitive is executed seems to be learned and embedded within the model itself**, rather than “in-context learning” from the prompt trajectory. This interpretation aligns with the Task-Representation (TR) hypothesis, if we think of an action primitive as a functional task.

Later in the _Appendix_, I noticed that ICRT demonstrates some ability to generalize to unseen but similar action primitives. Moreover, providing two or three prompt trajectories leads to consistently better performance compared to providing just one. This makes me wonder whether a mechanism similar to Task Learning (TL) is also at play in ICRT. In other words, could ICRT be performing a form of meta-learning, **learning how to execute action primitives based on demonstrations**?

This hypothesis could be further supported if we observe:
1. Consistent improvement as the number of prompt trajectories increases.
2. Successful generalization to significantly unfamiliar action primitives—those that differ substantially from the ones seen during training.

I'm wondering have you done such testing and what's your opinion on the role of prompt trajectory in ICRT? Do you think with much more demonstrations, the execution of action primitives will improve clearly?

Thanks for your patience and kindness,
Dong

**Comment by Max-Fu** [comment]
Hi, 

Based on the definition outlined in the seminal work of ICL, [GPT-3](https://arxiv.org/abs/2005.14165), ICL is equivalent to meta-learning without few-shot fine-tuning. By this definition, ICRT needs to perform meta-learning. For the scale that we tested ICRT, we don't see improvement in performance when the number of prompt trajectories increases, nor do we see a super clear signal on generalization to completely unseen motion primitives (see limitation/conclusion section of the paper, and the experimental results in appendix). We only see some signal on primitive compositions (i.e. if we combine two primitives together, sometimes there is such composition in eval). I think the key research question is how to correctly scale up ICRT (i.e. is L1 loss sufficient for learning motion, what is the correct vision, action, proprio representation, is the current ICRT formulation over compressing the data, etc.) so that we would have more GPT-3 like capabilities. 

Best,
Max

**Comment by MarSaKi** [comment]
Hi, thanks for sharing your insightful thoughts and I agree with your opinion. I'm going to close this issue 😊
